### PR TITLE
Feature/trace data collector

### LINKED ISSUE #45: Deserialize trace data files & fuse into one dataframe

### PR DESCRIPTION
**Pull request by RecurvedBow**
Adds a trace data collector class which searches for and deserializes trace data files in a given path. Can also delete the found trace data files so that old trace data files are not collected in another program run. 

Closes #45

### CODE REVIEW

File: tests/resource/sample_functions.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    for i, element1 in enumerate(list1):
+        element2 = list2[i]
+        are_elements_equal = sample_compare_integers(element1, element2)
+        if not are_elements_equal:
+            return False
+    return True
+
+
+def sample_compare_integers(value1, value2) -> bool:
```
Comment (bengsparks) [inline tests/resource/sample_functions.py:13]: did you mean to add the type hint here? or this even matter?

**Review by bengsparks — CHANGES_REQUESTED** [review]
You have added a binary test_pytype file to the repository; Please add this to the .gitignore and remove it.

**Comment by bengsparks** [comment]
Also, quite a few commits in here are just "pipeline fix"; We should squash these together prior to merging. I can show you how to do this when the time has come.

**Comment by RecurvedBow** [comment]
The reason I did this first is because I have to check whether the collected sample code can be successfully traced. What if the generated trace data files are empty/contain unexpected data? So to see whether the sample code is useful for the project, I can use the collector to view the data.

File: tests/resource/sample_functions.py (hunk 1 above)
Comment (RecurvedBow) [reply tests/resource/sample_functions.py]: It does not really matter, but I added them just for completeness.

File: tests/test_decorator_appender.py [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
 from tracing import TracerDecoratorAppender
 
-cwd = pathlib.Path.cwd() / 'resource' / "sample_decorator_appender_folder"
+cwd = pathlib.Path.cwd() / 'resource'
+
+
+def test_if_argument_of_append_decorator_is_none_error_is_raised():
+    test_object = TracerDecoratorAppender()
+    with pytest.raises(TypeError):
+        test_object.append_decorator_on_all_files_in(None)
```
[the remark this answers is not in the document]
Comment (RecurvedBow) [reply tests/test_decorator_appender.py]: Removed

File: tracing/decorator_appender.py [hunk 3]
```diff
@@ -17,6 +17,8 @@ def append_decorator_on_all_files_in(self, path: pathlib.Path,
                                          include_also_files_in_subdirectories: bool = False) -> None:
         """Finds all pytest files in the provided path argument and generates copies of these files with the
         decorators appended to the pytest functions. Stores the file paths of the new files."""
+        if path is None:
```
[the remark this answers is not in the document]
Comment (RecurvedBow) [reply tracing/decorator_appender.py]: Removed

File: tracing/tracer.py [hunk 4]
```diff
@@ -12,6 +12,9 @@
 
 class Tracer:
     def __init__(self, project_dir: pathlib.Path):
+        if project_dir is None:
```
[the remark this answers is not in the document]
Comment (RecurvedBow) [reply tracing/tracer.py:15]: Removed

**Comment by RecurvedBow** [comment]
> You have added a binary test_pytype file to the repository.
It is used for the collector tests, it is the expected data to compare with the actual (otherwise I would have to create a dataframe of over 50 rows).

>I also have doubts about deleting "old trace data files"

The deletion is optional; files won't be removed if the corresponding argument is False. I added it because the tests/resource folder was filled with pytype files and thought that adding this functionality won't hurt.

File: tests/resource/sample_functions.py (hunk 1 above)
Comment (bengsparks) [reply tests/resource/sample_functions.py]: Ok :)

**Review by bengsparks — CHANGES_REQUESTED** [review]
We need to come up with a different manner of handling the *_pytype files. They should not be stored in this repository.

**Comment by RecurvedBow** [comment]
> We need to come up with a different manner of handling the *_pytype files. They should not be stored in this repository.

To test the trace data file collector, sample trace data files are required (So that we can check if the collector collects these files). An alternate option is to run the strat to generate pytype files and clear them afterwards. However, this could cause the tests to be not unit tests anymore. This is why I decided to add these files into the test resource folder (The gitignore only makes an exception for the trace data files if they are in this folder, other files keep being ignored).

**Comment by bengsparks** [comment]
Could you put the test data into a separate repository and register that repository as a submodule that is stored in a third-party folder? Then change the tests to not run if these files are missing.

**Comment by RecurvedBow** [comment]
Is there any benefit in changing this to this option? Or rather, why is it bad to store *.pytype files in the repository? Because we probably need to store more trace data files for type unification or other data processing functions.

**Comment by bengsparks** [comment]
Binary artifacts are poor style in Git repositories. It is common for non-plain-text test resources to be stored somewhere externally.
The tests would not have to fetch the trace data files; they will be cloned when this PyType repositiory is cloned.

**Review by bengsparks — APPROVED** [review]

**Pull request merged**